### ISSUE TITLE
Use Edge Enhancement for the Dead Pixel Correction

### ISSUE THREAD

**Issue by kausthubhak** [10x-Engineers/Infinite-ISP#39]
In the current implementation, if the first pixel is bad, we wouldn't know that since we started testing the middle pixel of the next line. So, it's better to use edge enhancement techniques to include the first pixel in checking whether it's bad or not.

**Comment by marianadeem-10xe** [comment]
Yes, the current implementation assumes that all the neighboring pixels are good pixels and requires improvement. The case in which there are more than one bad pixel in a 3x3 window is not currently handled. However, the implementation does include a gradient based check to prevent edge blurring during DPC.

**Comment by wuyiulin** [comment]
Hi @marianadeem-10xe, I looked into this issue and found that the border problem is real, but it affects row/col **1** and **N-2**, not the first pixel.

### Root cause

`dynamic_dpc.py` relies on scipy's `mode="mirror"`, which reflects about the edge pixel. Same-color neighbors on a Bayer grid are 2 pixels apart, so for a pixel at index 1, the left neighbor (index -1) maps back to **the pixel itself**.

One row of an RGGB image:

```
index            -2   -1 |  0    1    2    3    4
raw row                  | R0   G0   R1   G1   R2

mode="mirror"    R1  [G0]| R0  [G0]  R1   G1   R2
                      ^-- left neighbor of G0 is G0 itself

CFA-aware pad    R1   G1 | R0  [G0]  R1   G1   R2
                      ^-- nearest real same-color pixel
```

Since the center pixel is in its own neighborhood, `img > max_value` can never hold and one of the differences is always 0, so detection always fails. The same happens at index N-2 on the other side, for both rows and columns.

### Minimal reproduction

```python
import numpy as np
from modules.dead_pixel_correction.dynamic_dpc import DynamicDPC

for pos in [(0, 0), (0, 1), (1, 10), (18, 10), (10, 10)]:
    img = np.full((20, 20), 100, np.float32)   # flat image
    img[pos] = 4000                            # one hot pixel
    out = DynamicDPC(img, {"height": 20, "width": 20, "bit_depth": 12},
                     {"dp_threshold": 80, "is_debug": False}).dynamic_dpc()
    print(pos, out[pos])
```

| position | region | current | with CFA-aware padding |
|---|---|---|---|
| (0, 0) | row/col 0 | 100 | 100 |
| (0, 1) | col 1 | **4000** | 100 |
| (1, 10) | row 1 | **4000** | 100 |
| (18, 10) | row N-2 | **4000** | 100 |
| (10, 10) | interior | 100 | 100 |

I also tested on `ColorChecker_2592x1536_12bits_RGGB.raw` with 2000 randomly injected singlets. Detection on rows/cols 1 and N-2 goes from 0% to 100%, and results for all other pixels are identical.

### Proposed fix

Pad each CFA channel separately before running the existing logic, then crop:

```python
H, W = img.shape
pad = np.empty((H + 4, W + 4), np.float32)
for py in (0, 1):
    for px in (0, 1):
        # reflect by 1 on each same-color sub-image = 2 pixels on the raw
        pad[py::2, px::2] = np.pad(img[py::2, px::2], 1, mode="reflect")

# ... run the existing detection/correction on `pad` ...

out = out[2:-2, 2:-2]
```

After this padding, every kept pixel's 5x5 window stays inside the padded array, so scipy's own `mode` no longer affects the result. The detection and correction logic itself doesn't need to change.

### Other notes

- `introduce_defect` in `util/utils.py` only places defects in rows/cols 2 to N-3, which may be why this wasn't caught earlier.
- As you mentioned, adjacent defects (couplets) are also not detected. I'd be happy to look into that separately.

Would you be open to a PR for the padding fix? Thanks!

**Comment by marianadeem-10xe** [comment]
Hi @wuyiulin, 

Thanks for digging into this. You're right about the boundary issue. I've reviewed your CFA-aware padding fix, please go ahead and generate a PR whenever you're ready (add a closing keyword referencing this issue in the PR description so it auto-closes on merge). 

You're welcome to take up the DPC improvement for adjacent defect pixel detection. Feel free to work on it, and reach out if you need any help.

Additionally, I noticed the same mirror-padding issue exists in Infinite-ISP_ReferenceModel's DPC module (the fixed-point model). I have opened a new issue there  as mentioned above. If you're interested in porting the fix over once this one's merged, that'd be great. Otherwise, it is open for anyone to work on.

Best,
Maria

**Comment by wuyiulin** [comment]
Thanks @marianadeem-10xe! I've opened #50 for the padding fix.

I'd be glad to look into the adjacent defect (couplet) detection after that, and I'm also interested in porting the fix to Infinite-ISP_ReferenceModel once this one is merged. I'll leave a comment on that issue as well.
